web/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql

```diff
@@ -0,0 +1,426 @@
+/*
+   Copyright 2021 Snowplow Analytics Ltd. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+
+/*
+  MK_TRANSACTION
+  Side effects procedure.
+  Input: a concatenation of one or more DML sql statements split by semicolon.
+  It is important that the statements are not DDL.
+  Either all the statements in the block succeed and get committed or all get rolled back.
+  To drop:
+  DROP PROCEDURE {{.output_schema}}.mk_transaction(VARCHAR);
+*/
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.mk_transaction(DML_STATEMENTS VARCHAR)
+  RETURNS VARCHAR
+  LANGUAGE JAVASCRIPT
+  AS
+  $$
+
+  dmls = DML_STATEMENTS.split(';')
+                       .filter(function(stmt) {return stmt.trim() !== '';})
+                       .map(function(stmt) {return stmt.trim() + ';';});
+
+  snowflake.createStatement({sqlText: `BEGIN;`}).execute();
+  try {
+
+      dmls.forEach(function(stmt) {
+          snowflake.createStatement({sqlText: stmt}).execute();
+      });
+      snowflake.createStatement({sqlText: `COMMIT;`}).execute();
+
+  } catch(ERROR) {
+      snowflake.createStatement({sqlText: `ROLLBACK;`}).execute();
+
+      // Snowflake error is not very helpful here
+      var err_msg = "Transaction rolled back.Probably failed: " + DML_STATEMENTS + " :Error: ";
+      throw Error(err_msg + ERROR);
+  }
+
+  return "ok. Statements in transaction succeeded."
+
+  $$
+;
+
+
+/*
+  COLUMN_CHECK
+  Checks for mismatched columns between source and target tables.
+  If source table is missing columns, it errors.
+  If AUTOMIGRATE is 'TRUE',
+    it allows the target table to get added the additional columns,if any, in source.
+  Since ALTER TABLE is a DDL statement, it will be its own transaction.
+  This means that even if it was placed inside another transaction,
+    it would commit that, and then start another(implicit) for its own execution.
+  So, also it cannot be explicitly rolled back.
+  This is why column_check is not part of the commit_with_metadata procedure.
```

Comment:
Could commit_table just contain two distinct transactions - one to check and add columns, and another to update the table?

Reply:
If i understand the question correctly, this is what essentially happens. The `ALTER` statement executes in its own transaction and commit does the updates. So the question is whether we want to wrap both inside a stored procedure. I chose not to, so as: a) to not create the illusion of a unit and b) `column_check` as a procedure of its own for reusability independent of committing. What do you think?

Reply:
What's an example of needing to use `column_check` independently of the commit table procedure?

Reply:
There isn't one here yet. The point of it being out of commit is that it cannot be done inside an explicit transaction, since the implicit one (enforced by `ALTER`, as by any DDL statement) would break the atomicity. (not sure if this answers)